### PR TITLE
feature selection with boruta algorithm

### PR DESCRIPTION
**Pull request by mmerce**
Please review whizzml code, but it would be great if an ML-expert could have a look too to check the algorithm.

### CODE REVIEW

File: boruta/script.whizzml [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+
+;; assoc+
+;;
+;; Adds the value in the key-value pair to the one existing in the accumulator
+;;
+;; Input:
+;;   acc: (map) accumulator map
+;;   key: (string) key to accumulate
+;;   value: (number) value to be added
+(define (assoc+ acc key value)
```
Comment (charleslparker) [inline boruta/script.whizzml:291]: I believe this is just `(assoc acc key (+ value (get acc key 0)))`  Correct me if I'm wrong, @jaor .

File: boruta/metadata.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "Boruta algorithm",
+  "description": "Finding important fields using a version of the Boruta algorithm",
+  "kind": "script",
+  "source_code": "script.whizzml",
+  "inputs": [
+      {
+          "name": "dataset-id",
+          "type": "dataset-id",
+          "description": "Select the dataset ID to be used"
```
Comment (jaor) [inline boruta/metadata.json]: just "Dataset to be analyzed"

File: boruta/readme.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+to detect important and unimportant fields in your dataset. The algorithm:
+
+- Uploads a data souce from its remote location
+- Creates the corresponding dataset
+- Creates a new extended dataset. In the new dataset, each field has a
+  corresponding shadow field which has the same type but contains a random
+  sample of the values contained in the original one.
+- Creates a random forest from the extended dataset
+- Extracts the maximum of the importances for the shadow fields
+- Uses this maximum plus (minus) a minim gain as threshold. Any of the original
```
Comment (jaor) [inline boruta/readme.md]: typo: minimum

File: boruta/readme.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  the remaining fields. The process stops when it reaches the user-given number
+  of runs or when all the original fields in the dataset are marked as
+  important or unimportant.
+
+The **inputs** for the script are:
+
+* `dataset-id`: (dataset-id) Dataset ID for the original data
+* `objective`: (string) Name of the objective field
+* `max-runs`: (integer) Maximum number of runs
+* `min-gain`: (number) Minimum amount of gain (real in the [0, 1] inteval).
```
Comment (jaor) [inline boruta/readme.md]: minimum gain

File: boruta/readme.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  Defines a neighbourhood
+  around the maximum importance of shadow fields. A field whose importance
+  falls in this interval will be considered "neutral" (neither important or
+  unimportant)
```
Comment (jaor) [inline boruta/readme.md:34]: typo: neither ... nor

File: boruta/script.whizzml [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+;;
+;; List of IDs of the fields in the dataset that can be chosen as objective
+;; field.
+;;
+;; Inputs:
+;;  fields: (map) Fields structure
+;; Output: (list) list of field IDs
+(define (choosable-objective-ids fields)
+  (let (field-val (lambda (fid k) (get-in fields [fid k]))
+        objective-types ["categorical", "numeric"]
```
Comment (jaor) [inline boruta/script.whizzml]: spurious ","

File: boruta/script.whizzml [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+;;
+;; Validates that the argument is a valid objective name in the reference
+;; dataset.
+;;
+;; Inputs:
+;;  objective-name: (string) Name of the objective field
+;;  dataset: (map) dataset resource information
+;;
+;; Output: (string) objective-name
+(define (dataset-objective-name? objective-name dataset)
```
Comment (jaor) [inline boruta/script.whizzml]: nit: this function is not a predicate (it doesn't return a boolean), so it's better not to name it with a ?... perhaps `validate-objective-name`

File: boruta/script.whizzml [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+;;
+;; Returns the name of the field used as objective field
+;;
+;; Inputs:
+;;  dataset: (map) Dataset resource info
+;;  objective-id: (string) ID of the objective field
+;;
+;; Outputs: (string) Name of the objective field
+
+(define (get-objective-name dataset objective-id)
```
Comment (jaor) [inline boruta/script.whizzml]: nit: this function just get the name of a field, it doesn't care whether it's an objective or not... perhaps 
`(get-field-name dataset field-id)` is a better signature.

File: boruta/script.whizzml [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+        field-ids (keys ori-fields)
+        f-names (map (lambda (x) (get-in ori-fields [x "name"])) field-ids)
+        init-rank (reduce (lambda (x y) (assoc x (get-in ori-fields
+                                                         [y "name"]) 0))
+                          {}
+                          field-ids)
+        objective (if (= objective "")
+                      (get-objective-name ds
+                                          (dataset-get-objective-id
+                                           dataset-id))
```
Comment (jaor) [inline boruta/script.whizzml]: indentation: two spaces to the right

File: boruta/script.whizzml [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+                                                                to-names))
+                _ (log-info "Unimportant: " (select-by-importance f-rank
+                                                                  IMPORTANCE_DEC
+                                                                  to-names))
+                _ (log-info "Neutral: " (select-by-importance f-rank
+                                                              0
+                                                              to-names))
+                fields (remove-unimportant fields f-rank)
+                _ (log-info "remaining fields: " (map to-names (keys fields)))
+                _ (log-info "------------------------------------------------"))
```
Comment (jaor) [inline boruta/script.whizzml]: you can put all these logs in the body of the `let`, right before he `recur`, without having to assign them to _ and with a bit more space :)

File: boruta/script.whizzml [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+        neutral (map to-names
+                     (keys (filter-map (lambda(x y) (= y 0))
+                                       f-rank)))
+        unimportant (filter (lambda (x) (not (or (member? x important)
+                                                 (member? x neutral))))
+                            f-names))
+    (assoc (assoc {}
+                  "important" important)
+                  "unimportant" unimportant
+                  "neutral" neutral)))
```
Comment (jaor) [inline boruta/script.whizzml]: indentation is off, and you can write it without any assoc:

```
{"important" important
 "unimportant" unimportant
 "neutral" neutral}
```

File: boruta/script.whizzml [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+;;   fields: (map) New fields structure for the fields in the original dataset.
+;;                 This structure is needed to know the IDs in the
+;;                 shadow dataset for the original fields
+;;
+(define (shadow-ds ds-id fields obj-name)
+  (let (f-id-names (reduce (lambda (x y) (assoc x
+                                                y
+                                                (get-in fields [y "name"])))
+                           {}
+                           (keys fields))
```
Comment (jaor) [inline boruta/script.whizzml:220]: formatting nit: it's usual to put lambda's bodies below the parameter list, just as in defines, and one gets a bit more room: (and, at least to my eyes, it reads a bit better):

```
(let (f-id-names (reduce (lambda (x y)
                           (assoc x y (get-in fields [y "name"])))
                         {}
                         (keys fields))
      obj-f (filter-map (lambda (x _)
                          (= (get-in fields [x "name"]) obj-name))
                        f-id-names)
```

File: boruta/script.whizzml [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+                                           "new_fields"
+                                           (map new-fields flatline-fs)})
+        sh-ds (fetch sh-ds-id)
+        sh-fields (get sh-ds "fields")
+        sh-f-ids (filter (lambda (x) (member? (get-in sh-fields [x "name"])
+                                              f-names))
+                         (keys sh-fields))
+        old-fields (filter-map (lambda (x _) (not (member? (get-in sh-fields
+                                                           [x "name"])
+                                                  f-names)))
```
Comment (jaor) [inline boruta/script.whizzml]: indentation is off:

```
(not (member? (get-in sh-fields
                      [x "name"])
              f-names))
```

or with the recommended lambda body (and, optionally, up to you) a let:

```
old-fields (filter-map (lambda (x _)
                         (let (n (get-in sh-fields [x "name"]))
                           (not (member? n f-names))))
                       sh-fields)
```

File: boruta/script.whizzml (hunk 1 above)
Comment (mmerce) [inline boruta/script.whizzml:291]: yes! much nicer!

File: boruta/script.whizzml [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+                               sh-fields))
+    [sh-ds-id sh-f-ids old-fields]))
+
+;; repeat-fields
+;; Repeats the field ID until the number of elements is the one chosen
```
Comment (jaor) [inline boruta/script.whizzml]: Cyclically repeats the list of field IDs until it has the desired number of elements

File: boruta/script.whizzml (hunk 1 above)
Comment (jaor) [inline boruta/script.whizzml:291]: yes, that's correct

File: boruta/script.whizzml [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+;;
+;; Outputs: list of importances by field name
+
+(define (get-importances ensemble)
+  (let (dist (get ensemble "distributions")
+        imps (map (lambda (x) (get x "importance")) dist)
+        m-num (count imps))
+    (iterate (acc {} m-imps imps)
+      (iterate (m-acc acc imp m-imps)
+        (let (f-id (head imp))
```
Comment (jaor) [inline boruta/script.whizzml]: you don't use `f-id` in the body...

File: boruta/script.whizzml [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+;; Selects the importance information for each of the fields in the ensemble
+;;
+;; Inputs:
+;;   ens-id: (ensemble-id) Ensemble ID
+;;
+;; Output: (map) field-name - importance map
+
+(define (importances ens-id)
+  (let (ens (fetch ens-id))
+    (get-importances ens)))
```
Comment (jaor) [inline boruta/script.whizzml]: Perhaps: `(get-importances (fetch ens-id))`

File: boruta/script.whizzml [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+;;
+;; Inputs:
+;;   imps: (list) List of field importances
+;;   sh-ids: (list) List of field IDs of the shadow fields
+;;
+;; Output: (number) maximum importance for the fields not in the list
+(define (maximum-shadow-importance imps sh-ids)
+  (iterate (acc 0 imp-k (keys imps))
+    (if (member? imp-k sh-ids)
+      (max [acc (get imps imp-k)])
```
Comment (jaor) [inline boruta/script.whizzml]: `(max acc (get imps imp-k))`

File: boruta/test/test.sh [hunk 18]
```diff
@@ -0,0 +1,36 @@
+VERBOSITY=0
+rm -f -R cmd_del
+rm -f -R cmd
+rm -f -R .build
+echo "-------------------------------------------------------"
+echo "Test for boruta script"
+bigmler whizzml --package-dir ../ --output-dir ./.build --verbosity $VERBOSITY
+# creating the resources needed to run the test
+bigmler --train s3://bigml-public/csv/iris.csv --no-model --project "Whizzml examples tests" --output-dir cmd/pre_test --verbosity $VERBOSITY
```
Comment (jaor) [inline boruta/test/test.sh]:
``` bash
bigmler --train s3:/....  \
        --no-model \
        --project "whizzml ..." \
        ...
```

File: cross-validation/test/test.sh [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+prefix='[["dataset-id", "'
+suffix='"]]'
+text=''
+cat cmd/pre_test/dataset | while read dataset
+do
+echo "$prefix$dataset$suffix" > "test_inputs.json"
+done
+echo "Testing basic script ----------------------------------"
+# running the execution with the given inputs
+bigmler execute --scripts .build/basic/scripts --inputs test_inputs.json --output-dir cmd/results --verbosity $VERBOSITY
```
Comment (jaor) [inline cross-validation/test/test.sh]: bigmler with `\` as above

File: boruta/script.whizzml (hunk 8 above)
Comment (mmerce) [inline boruta/script.whizzml]: I'll remove it. We can simply use get-in from ori-fields

**Pull request merged**